kapitan/targets.py

```diff
@@ -70,11 +74,22 @@ def compile_targets(inventory_path, search_paths, output_path, parallel, targets
 
         # append "compiled" to output_path so we can safely overwrite it
         compile_path = os.path.join(output_path, "compiled")
-        worker = partial(compile_target, search_paths=search_paths, compile_path=temp_path, ref_controller=ref_controller,
+        worker = partial(compile_target, search_paths=search_paths, compile_path=temp_path,ref_controller=ref_controller,
                          **kwargs)
 
-        if target_objs == []:
+        if not target_objs:
             raise CompileError("Error: no targets found")
+
+        fetch = partial(fetch_dependencies, cache_path=kwargs.get('dependency_cache_path', None))
+        target_dependencies_map = get_dependencies_per_target(target_objs)
+        if kwargs.get('force_fetch', False):
+            # force fetch
+            remove_saved_dependencies(target_dependencies_map, kwargs.get('dependency_cache_path', None))
+
+        for target_name, items in target_dependencies_map.items():
+            [p.get() for p in pool.imap_unordered(fetch, items) if p]
+            logger.info("Dependency fetching complete for target {}".format(target_name))
```

Comment:
Fetching is done concurrently among dependencies, target by target. I think it makes sense because two different targets could contain the same (inherited) dependencies, in which case we could be fetching the same thing at the same time unless we first take care of such duplicates. If we do them target-by-target, then the second one onward could simply check if the file already exists inside fetch() function and avoid fetching twice.

Reply:
Have you tested this with `kapitan compile -p 4` for example? With 4 parallel threads? I'm curious how it would behave then when multiple targets re-use a dependency. :)

Reply:
@uberspot when multiple targets use the same dependency, then since fetching is done target-by-target sequentially, the set of dependencies for the first target will be fetched and then if there are any duplicates for the remaining targets, then the fetch would be skipped. This is done by checking whether the file at the output_path exists for http types, and for git types, whether the repository exists in cache.

Reply:
The potential issue is when a target uses the same dependency (source) for some reason. Then there could be situations where for example git clone is done simultaneously for the same repo, and one clone process finishes saving to output_path, and the other clone process throws error because the output_path already exists.

Reply:
In that case ^ we don't want to error out and stop the compilation. The user can't do anything to solve this ^. So kapitan should just ideally ignore it and assume the dependency is already downloaded. And add logger.debug() messages to make it easy to detect if that happens when debugging. :+1:

Reply:
@uberspot Makes sense! I'll catch the exception and log that message.

Reply:
This makes me think that perhaps compilation shouldn't be downloading dependencies and maybe we should introduce a new kapitan subcommand for deps? (just a thought)

Reply:
@ramaro Possibly we can do both? Allow the user to specify flag for `compile` command that does fetching before compile. Since fetching is not per-target, I'm actually starting to see it as a different process to compile as well.

Reply:
Isn't that the point of `compile --fetch`?

Reply:
@ramaro ok I see. I originally thought `--fetch` was to force the fresh dependency fetch (to update the existing files when the external resources are updated), where fetching was done automatically prior to compile.

Reply:
That's what i thought as well. It would be simpler for users if fetching was seamless, at least for the first time (?)

Reply:
Should `fetch` now be a separate command as well, with `--fetch` a flag for `compile`?